klippa/src/graph.rs

```diff
@@ -0,0 +1,515 @@
+            for link in next_obj
+                .real_links
+                .values()
+                .chain(next_obj.virtual_links.values())
+            {
+                let child_idx = link.obj_idx();
+                removed_edges[child_idx] += 1;
+
+                let child_v = &self.vertices[child_idx];
+                if child_v.incoming_edges() == removed_edges[child_idx] {
+                    queue.push((child_v.modified_distance(order), child_idx));
+                    order += 1;
+                }
+            }
+        }
+
+        std::mem::swap(&mut self.ordering, new_ordering);
+        if pos != v_count {
+            self.set_err();
+            //TODO: add print_orphaned_nodes()
+        }
+    }
+
+    fn root(&self) -> &Vertex {
+        &self.vertices[self.root_idx()]
+    }
+
+    pub(crate) fn update_parents(&mut self) {
+        if !self.parents_invalid {
+            return;
+        }
+
+        for v in self.vertices.iter_mut() {
+            v.reset_parents();
+        }
+
+        let count = self.vertices.len();
+        let mut real_links_idxes = Vec::with_capacity(count);
+        let mut virtual_links_idxes = Vec::with_capacity(count);
+        for idx in 0..count {
+            let obj = &self.vertices[idx].obj;
+
+            real_links_idxes.clear();
+            virtual_links_idxes.clear();
+            for l in obj.real_links.values() {
+                real_links_idxes.push(l.obj_idx());
+            }
+
+            for l in obj.virtual_links.values() {
+                virtual_links_idxes.push(l.obj_idx());
+            }
+
+            for child_idx in &real_links_idxes {
+                let Some(v) = self.vertices.get_mut(*child_idx) else {
+                    self.set_err();
+                    return;
+                };
+                v.add_parent(idx, false);
+            }
+
+            for child_idx in &virtual_links_idxes {
+                let Some(v) = self.vertices.get_mut(*child_idx) else {
+                    self.set_err();
+                    return;
+                };
+                v.add_parent(idx, true);
+            }
+        }
+    }
+
+    // Finds the distance to each object in the graph from the root node
+    // Uses Dijkstra's algorithm to find all of the shortest distances.
+    // ref: <https://github.com/harfbuzz/harfbuzz/blob/3f70b6987830bba1e4922cad03028cdd9d78b3a1/src/graph/graph.hh#L1512C13-L1512C72>
+    pub(crate) fn update_distances(&mut self) {
+        if !self.distance_invalid {
+            return;
+        }
+
+        for i in self.vertices.iter_mut() {
+            i.distance = u64::MAX / 2;
+        }
+        let root_idx = self.root_idx();
+        self.vertices[root_idx].distance = 0;
+
+        let count = self.vertices.len();
+        let mut queue = PriorityQueue::with_capacity(count);
+        queue.push((0_u64, root_idx));
+
+        let mut visited = vec![false; count];
+        let mut distance_map = FnvHashMap::default();
```

Comment:
Since key is an object index and we expect one value per vertex, you can use a presized vector as a map which will be a little bit faster.